stdlib/experimental/prometheus/from.go

```diff
@@ -0,0 +1,339 @@
+type Metric struct {
+	Name      string                 `json:"name"`
+	Tags      map[string]string      `json:"tags"`
+	Fields    map[string]interface{} `json:"fields"`
+	Timestamp time.Time              `json:"timestamp"`
+	Type      string                 `json:"string"` // need to add other funcs for validation; first make sure this is right
+}
+
+func (p *PrometheusIterator) Connect(ctx context.Context) error {
+
+	return nil
+}
+
+func (p *PrometheusIterator) parse(r io.Reader, header http.Header) (err error) {
+	var parser expfmt.TextParser
+	now := time.Now()
+
+	mediatype, params, err := mime.ParseMediaType(header.Get("Content-Type"))
+	if err != nil {
+		return err
+	} // why do we need to parse the media type?
+
+	// Prepare output
+	metricFamilies := make(map[string]*dto.MetricFamily)
+	if mediatype == "application/vnd.google.protobuf" &&
+		// read about protobufs
+		params["encoding"] == "delimited" &&
+		params["proto"] == "io.prometheus.client.MetricFamily" {
+		for {
+			mf := &dto.MetricFamily{}
+			if _, err := pbutil.ReadDelimited(r, mf); err != nil {
+				if err == io.EOF {
+					break
+				}
+				return fmt.Errorf("reading metric family protocol buffer failed: %s", err)
+			}
+			metricFamilies[mf.GetName()] = mf
+		}
+	} else {
+		metricFamilies, err = parser.TextToMetricFamilies(r)
+		if err != nil {
+			return fmt.Errorf("reading text format failed: %s", err)
+		}
+	}
+
+	p.metrics = make([]Metric, 0)
+
+	// read metrics
+	for name, family := range metricFamilies {
+		for _, m := range family.Metric {
+			// read tags
+			tags := makeLabels(m)
+
+			// read fields
+			fields := make(map[string]interface{})
+
+			switch family.GetType() {
+			case dto.MetricType_SUMMARY:
+				continue // skip for now
+			case dto.MetricType_HISTOGRAM:
+				continue // skip for now
+			default: // counters, gagues and uptyped
+				fields = getNameAndValue(m)
+			}
+
+			// convert to row for flux Table
+			if len(fields) > 0 {
+				tm := now
+				if m.TimestampMs != nil && *m.TimestampMs > 0 {
+					tm = time.Unix(0, *m.TimestampMs*1000000)
+				}
+				met := Metric{
+					Timestamp: tm,
+					Tags:      tags,
+					Fields:    fields,
+					Name:      name,
+					// Type:      string(family.GetType()),
+				}
+				p.metrics = append(p.metrics, met)
+			}
+
+		}
+	}
+
+	return nil
+}
+
+func (p *PrometheusIterator) Fetch(ctx context.Context) (bool, error) {
+	p.url = "http://demo.robustperception.io:9090/metrics"
+	resp, err := http.Get(p.url)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+
+	// parse the entire body into a slice of Metrics
+	err = p.parse(resp.Body, resp.Header)
+	if err != nil {
+		return false, err
+	}
+
+	// Iterate over all Metrics in the slice
+	if p.i < len(p.metrics) {
+		p.nextMetric()
+		return true, nil
+	}
+	// p.met = p.metrics[0]
+	return false, nil
+}
+
+func (p *PrometheusIterator) nextMetric() error {
+	fmt.Println(p.i)
+	next := p.metrics[p.i]
```

Comment:
this seems like a lot of unnecessary copying.  Not sure if this is accountable for the slowdown, but just advance `p.i` here (or just drop this fn and advance `i` on line 220. )

then, in `Decode` just use `p.metrics[p.i]` in the few places it's needed

Reply:
Is the assignment the thing that is slow here? Or is it indexing the slice?

Reply:
I thought both of those should be O(1). Sorry if your point went over my head here.

Reply:
Nevermind. Chris explained the issue here. I'll go ahead and fix it 👍🏾